include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,453 @@
+
+            // Check if the projected segment intersects the polygon
+            std::tuple
+            <
+                boost::geometry::model::multi_point<point_2d>,
+                boost::geometry::model::multi_linestring<linestring_2d>> intersections;
+            boost::geometry::intersection(projected_segment, projected_polygon, intersections);
+
+            auto& points = std::get<0>(intersections);
+            auto& linestrings = std::get<1>(intersections);
+
+            if (boost::geometry::is_empty(points) && boost::geometry::is_empty(linestrings))
+            {
+                return intersection_type::none; // No intersection, this is valid
+            }
+            if (!points.empty())
+            {
+                visitor.template apply<failure_invalid_intersection>();
+                return intersection_type::invalid;
+            }
+            if (linestrings.size() == 1)
+            {
+                if (boost::geometry::equals(linestrings.front(), projected_segment))
+                {
+                    if (!has_correct_intersection_orientation(projected_polygon, projected_p1,
+                                                              projected_p2, visitor))
+                    {
+                        return intersection_type::invalid;
+                    }
+                    return intersection_type::valid; // The only case of valid non empty intersection
+                }
+            }
+            visitor.template apply<failure_invalid_intersection>();
+            return intersection_type::invalid;
+        }
+
+        // Compute the intersection point
+        vec3 seg_dir = p2 - p1;
+        CT denom = dot(normal, seg_dir);
+        assert(denom != 0); // Ensure the segment is not parallel to the plane
+        CT t = dot(normal, base - p1) / denom;
+        vec3 intersection = p1 + t * seg_dir;
+
+        // Project the intersection point onto the plane of the polygon
+        auto projected_intersection = project_point_to_plane_2d(v1, v2, intersection);
+
+        // Check if the projected intersection point is inside the projected polygon
+        // If the segment has one points on the plane then we consider the intersection with the boundary
+        // as a an empty instersection i.e. not invalid but also insufficient for considering
+        // connectivity
+        bool intersects = std::abs(sign1) > tolerance && std::abs(sign2) > tolerance ?
+            boost::geometry::covered_by(projected_intersection, projected_polygon, strategy) :
+            boost::geometry::within(projected_intersection, projected_polygon, strategy);
+        if (intersects)
+        {
+            visitor.template apply<failure_invalid_intersection>();
+            return intersection_type::invalid;
+        }
+        return intersection_type::none;
+    }
+
+    // Helper function to check if any segment of a ring intersects a polygon
```

Comment:
Do we really support polygons? Or rings? It's not completely clear to me.
This is checking the outer ring only?